### PR TITLE
Add dif cam1 2 3

### PR DESCRIPTION
**Pull request by ambarb**
this was adding hdf5 saving ability with ROIstats, but not turning hdf5 saving on/off for 1 device.  The idea is to switch between devices.

did the same for tiff saving of color cameras (dif_cam2 and dif_cam1).  Outstanding issue with trans in dif_cam3, but we can fix this as the cycle goes along.

also found roi configuration attrs were not being saved for fccd .  fixed that and added also for dif_beam.  this could be done better, but not before the cycle starts tomorrow.

### CODE REVIEW

File: startup/csx1/startup/detectors.py [hunk 1]
```diff
@@ -113,12 +134,20 @@ def _setup_stats(cam_in):
...
+##TODO make roi config attrs into generic function like _setup_stats so all areadetectors can have roi coordinates
 for attr in configuration_attrs_list:
     getattr(fccd, attr).kind='config'
+fccd.configuration_attrs.extend(['roi1', 'roi2', 'roi3','roi4'])
+_setup_stats(fccd)
 
 
-_setup_stats(fccd)
+configuration_attrs_list = []                        
+configuration_attrs_list.extend(['roi' + str(i) + string for i in range(1,5) for string in roi_params])
```
Comment (mrakitin) [inline startup/csx1/startup/detectors.py:150]: What's an example of `roi_params`?

**Review by mrakitin — COMMENTED** [review]
Looks good overall. Suggested a few minor improvements.

File: startup/csx1/devices/areadetector.py [hunk 2]
```diff
@@ -36,24 +36,96 @@ class StandardCam(SingleTrigger, AreaDetector):
...
-        return self.stast1.cenx.subscribe(*args, **kwargs)
+        return self.stast1.cenx.subscribe(*args, **kwargs) #TODO if this works, then add stats1.ceny too.
 
     def unbuscribe(self, *args, **kwargs):
         return self.stast1.cenx.unsubscribe(*args, **kwargs)
 
 
+class HDF5PluginWithFileStorePlain(HDF5Plugin, FileStoreHDF5IterativeWrite): ##SOURCED FROM BELOW FROM FCCD WITH SWMR removed
+    # AD v2.2.0 (at least) does not have this. It is present in v1.9.1.
+    file_number_sync = None
```
[the remark this answers is not in the document]
Comment (maffettone) [reply startup/csx1/devices/areadetector.py]: https://github.com/NSLS-II-CSX/profile_collection/pull/66/commits/d48455b2dd63d87d863e67fc32e7b6ea0a681a25

File: startup/csx1/devices/areadetector.py [hunk 3]
```diff
@@ -36,24 +36,96 @@ class StandardCam(SingleTrigger, AreaDetector):
...
+            ret[key]['shape'] = [
+                self.parent.cam.num_images.get(),
+                self.array_size.height.get(),
+                self.array_size.width.get()
+                ]
+       
+        elif color_mode in ['RGB1', 'Bayer']:
+            ret[key]['shape'] = [self.parent.cam.num_images.get(), *self.array_size.get()]
+        else:
+            raise RuntimeError("SHould never be here")
```
[the remark this answers is not in the document]
Comment (maffettone) [reply startup/csx1/devices/areadetector.py]: https://github.com/NSLS-II-CSX/profile_collection/pull/66/commits/d48455b2dd63d87d863e67fc32e7b6ea0a681a25

File: startup/csx1/devices/areadetector.py [hunk 4]
```diff
@@ -36,24 +36,96 @@ class StandardCam(SingleTrigger, AreaDetector):
...
+        
+        cam_dtype = self.parent.cam.data_type.get(as_string=True)
+        type_map = {'UInt8': '|u1', 'UInt16': '<u2', 'Float32':'<f4', "Float64":'<f8'}
```
[the remark this answers is not in the document]
Comment (maffettone) [reply startup/csx1/devices/areadetector.py]: https://github.com/NSLS-II-CSX/profile_collection/pull/66/commits/d48455b2dd63d87d863e67fc32e7b6ea0a681a25

File: startup/csx1/devices/areadetector.py [hunk 5]
```diff
@@ -36,24 +36,96 @@ class StandardCam(SingleTrigger, AreaDetector):
...
+        if cam_dtype in type_map:
+            ret[key].setdefault('dtype_str', type_map[cam_dtype])
+        return ret
+
+
+class StandardProsilicaWithHDF5(StandardCam):
+    hdf5 = Cpt(HDF5PluginWithFileStorePlain,
+              suffix='HDF1:',
+              write_path_template='/nsls2/data/csx/legacy/prosilica_data/%Y/%m/%d',
```
[the remark this answers is not in the document]
Comment (maffettone) [reply startup/csx1/devices/areadetector.py]: This needs a follow up to ensure these directories exist and can be written to by the detector. 
[3c14622](https://github.com/NSLS-II-CSX/profile_collection/pull/66/commits/3c14622c6a283ffa411fe82f97c47b91f31139fe)

**Comment by maffettone** [comment]
@mrakitin 
I made edits to ensure that the HDF5 plugin will be able to create directories. 
https://github.com/NSLS-II-CSX/profile_collection/pull/66/commits/f909ee7a5b2f6e5fa2ee0dcc051eac8068ed3840

This assumes that even though https://github.com/bluesky/ophyd/pull/996 is up to date in the environment, that a staging order will cause issues in the HDF5 plugin (https://github.com/bluesky/ophyd/issues/1027). 
Following the example from https://github.com/NSLS-II-SIX/profile_collection/pull/36/files.

File: startup/csx1/devices/areadetector.py [hunk 6]
```diff
@@ -36,24 +36,108 @@ class StandardCam(SingleTrigger, AreaDetector):
...
+        dic[key].setdefault('dtype_str', type_map[cam_dtype])
+
+
+class HDF5PluginWithFileStorePlain(HDF5Plugin_V22, FileStoreHDF5IterativeWrite): ##SOURCED FROM BELOW FROM FCCD WITH SWMR removed
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # In CSS help: "N < 0: Up to abs(N) new directory levels will be created"
+        self.stage_sigs.update({"create_directory": -3})
+        self.stage_sigs.move_to_end("create_directory", last=False)
```
Comment (mrakitin) [inline startup/csx1/devices/areadetector.py:82]: I like this solution, I hope it works as expected :)

On a side note, `.move_to_end(..., last=False)` is the worst API 😁. Good it's explained here: https://docs.python.org/3/library/collections.html#collections.OrderedDict.move_to_end. Maybe worth adding a comment above that line that we are moving the key to the beginning of the dict?

Comment (maffettone) [reply startup/csx1/devices/areadetector.py]: Indeed, we'll test out today. I've added a note in https://github.com/NSLS-II-CSX/profile_collection/pull/66/commits/434e050d807aa52ce5e85ce067a64d029dde6b5b. I was definitely taking the convoluted logic for granted...

**Review by maffettone — COMMENTED** [review]
After forcing `self.stage_sigs.update({"create_directory": -3})` for both the TIFF and HDF5 components, and fixing some bugs in the `describe` method, these changes are ready to merge.

**Pull request merged**